File: 백준/Silver/15657. N과 M （8）/N과 M （8）.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int n,m;
vector<int> arr;
int path[9];
int visited[9]={0,};
void dfs(int lev, int num){
  if(lev==m){
    for(int i=0;i<m;i++){
      cout<<path[i]<<" ";
    }
    cout<<'\n';
    return ;
  }
  for(int i=num;i<n;i++){
    path[lev]=arr[i];
    dfs(lev+1,i);
  }
}
int main(){
  cin.tie(nullptr);
  cout.tie(nullptr);
  cin>>n>>m;
  for(int i=0;i<n;i++){
    int num;
    cin>>num;
    arr.push_back(num);
  }
  sort(arr.begin(), arr.end());
  dfs(0,0);

}

File: 백준/Silver/15657. N과 M （8）/N과 M （8）.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
	public static StringBuilder sb = new StringBuilder();
	public static int[] arr;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] tmp = br.readLine().split(" ");
		int n = Integer.parseInt(tmp[0]);
		int m = Integer.parseInt(tmp[1]);
		String[] docs = br.readLine().split(" ");
		arr = new int[n];
		int[] path = new int[n];
		for(int i=0;i<n;i++) {
			arr[i] = Integer.parseInt(docs[i]);
		}
		Arrays.sort(arr);
		dfs(0,0,m, path);
		System.out.println(sb.toString());
	}
	
	public static void dfs(int cri, int lev, int level, int[] path) {
		if(lev==level) {
			for(int i=0;i<lev;i++) {
				sb.append(path[i]).append(" ");
			}
			sb.append("\n");
			return;
		}
		for(int i=cri;i<arr.length;i++) {
			path[lev] = arr[i];
			dfs(i, lev+1, level, path);
		}
	}
}
